### DISCUSSION TITLE
dns resolve ipv6.google.com - ResponseTimeout

### DISCUSSION THREAD

**Discussion by abu-matterize** [openthread/openthread#7777] [Q&A]
Hello,

When I try `dns resolve ipv4.google.com` it timeouts.
```
> dns resolve ipv4.google.com
> DNS response for ipv4.google.com - Error 28: ResponseTimeout
```

[TL;DR]

I cloned the `ot-br-posix` https://github.com/openthread/ot-br-posix and checkout to `4b04548` as mentioned in https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_thread_tools.html and build with change in `_otbr` as mentioned in https://groups.google.com/g/openthread-users/c/OdnFWe08yVg/m/jJQ8PY1pCQAJ
```
"-DOTBR_DNSSD_DISCOVERY_PROXY=OFF"
"-DOTBR_SRP_ADVERTISING_PROXY=OFF"
```

My border router config:
```
OS: Raspbian GNU/Linux 10 (buster) armv7l 
Host: Raspberry Pi 4 Model B Rev 1.4 
Kernel: 5.10.103-v7l+ 
Shell: bash 5.0.3 
Terminal: /dev/pts/0 
CPU: BCM2711 (4) @ 1.500GHz 
Memory: 87MiB / 3838MiB
```
I've built `rcp` from source https://github.com/openthread/ot-nrf528xx with default config for nRF52840 Dongle (PCA10059) and plugged the dongle to my border router.

When I run `ot-ctl` it enters the CLI `>`. The `otbr-agent` defaults is
```
pi@router:~ $ cat /etc/default/otbr-agent 
# Default settings for otbr-agent. This file is sourced by systemd

# Options to pass to otbr-agent
OTBR_AGENT_OPTS="-I wpan0 -B wlan0 spinel+hdlc+uart:///dev/ttyACM0 trel://wlan0"
OTBR_NO_AUTO_ATTACH=0
pi@router:~ $
```
When I `ping 64:ff9b::808:808` from another FTD in the network it responds.
```
> ping 64:ff9b::808:808
Done
> 16 bytes from 64:ff9b:0:0:0:0:808:808: icmp_seq=3 hlim=116 time=32ms

```
Border router `netdata show` is
```
pi@router:~ $ sudo ot-ctl netdata show
Prefixes:
fde8:1a47:2eb3:1::/64 paos med 1800
Routes:
2409:4072:6d91:5896::/64 s med 1800
64:ff9b:0:0::/64 med 1800
Services:
44970 01 7a04b000000e10 s 1800
Done
pi@router:~ $ 
```

But `dns resolve ipv6.google.com` fails
```
> dns resolve ipv4.google.com
> DNS response for ipv4.google.com - Error 28: ResponseTimeout
```

Border Router Interface: `ip addr` dump
```
pi@router:~ $ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether e4:5f:01:0d:84:c2 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.18/24 brd 192.168.0.255 scope global dynamic noprefixroute eth0
       valid_lft 83207sec preferred_lft 72407sec
    inet6 fe80::e65f:1ff:fe0d:84c2/64 scope link 
       valid_lft forever preferred_lft forever
3: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether e4:5f:01:0d:84:c3 brd ff:ff:ff:ff:ff:ff
    inet 192.168.43.32/24 brd 192.168.43.255 scope global dynamic noprefixroute wlan0
       valid_lft 2229sec preferred_lft 1664sec
    inet6 2409:4072:6d91:5896:e65f:1ff:fe0d:84c3/64 scope global dynamic mngtmpaddr 
       valid_lft 3533sec preferred_lft 3533sec
    inet6 fe80::e65f:1ff:fe0d:84c3/64 scope link 
       valid_lft forever preferred_lft forever
4: wpan0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1280 qdisc pfifo_fast state UNKNOWN group default qlen 500
    link/void 
    inet6 fde8:1a47:2eb3:1:e61e:f24e:8f8b:d637/64 scope global nodad 
       valid_lft forever preferred_lft forever
    inet6 fd2a:a5fa:d26e:1e86:0:ff:fe00:fc10/64 scope global nodad deprecated 
       valid_lft forever preferred_lft 0sec
    inet6 fd2a:a5fa:d26e:1e86:0:ff:fe00:fc38/64 scope global nodad deprecated 
       valid_lft forever preferred_lft 0sec
    inet6 fd2a:a5fa:d26e:1e86:0:ff:fe00:1800/64 scope global nodad deprecated 
       valid_lft forever preferred_lft 0sec
    inet6 fd2a:a5fa:d26e:1e86:cefe:f40d:59cc:b681/64 scope global nodad deprecated 
       valid_lft forever preferred_lft 0sec
    inet6 fe80::f862:f7a6:bc8d:805d/64 scope link nodad 
       valid_lft forever preferred_lft forever
5: nat64: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 500
    link/none 
    inet 192.168.255.1/32 scope global nat64
       valid_lft forever preferred_lft forever
    inet6 fdaa:bb:1::2/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::cf5c:5578:a9cf:d16/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
pi@router:~ $ 

```
`ot-ctl ipaddr` dump
```
pi@router:~ $ sudo ot-ctl ipaddr
fde8:1a47:2eb3:1:e61e:f24e:8f8b:d637
fd2a:a5fa:d26e:1e86:0:ff:fe00:fc10
fd2a:a5fa:d26e:1e86:0:ff:fe00:fc38
fd2a:a5fa:d26e:1e86:0:ff:fe00:1800
fd2a:a5fa:d26e:1e86:cefe:f40d:59cc:b681
fe80:0:0:0:f862:f7a6:bc8d:805d
Done
pi@router:~ $ 
```
CLI Node `ipaddr` dump
```
> ipaddr
fde8:1a47:2eb3:1:4b0b:e25a:65a4:afbc
fd2a:a5fa:d26e:1e86:0:ff:fe00:fc00
fd2a:a5fa:d26e:1e86:0:ff:fe00:7400
fd2a:a5fa:d26e:1e86:9fd7:5733:fc7:61b7
fe80:0:0:0:1ce2:e70a:7746:5ceb
Done
> 
```
What I'm missing in the configuration?

**Comment by jwhui** [comment]
I believe the DNS client is not configured with an appropriate DNS server address.

One thing you can try is by enabling the SRP auto-configure mode, which also auto-configures the DNS client:
```
srp client autostart enable
```
If successful, you should see the DNS config provide the IPv6 address of a Thread Border Router.
```
dns config
```

**Reply by abu-matterize** [reply]
@jwhui - Thanks for pointing. My CLI node was built with `nRF5_SDK_for_Thread_and_Zigbee_v4.2.0_af27f76` and it throws an invalid command for `srp` and `dns config`
```
> srp client autostart enable
Error 35: InvalidCommand
> dns config
Error 35: InvalidCommand
> 
```
I tried build from https://github.com/openthread/ot-nrf528xx/tree/main/src/nrf52833 and the build is successful but the result for the `srp` and `dns config` is same as above.

Then I switched to nRF Connect SDK v1.9.1 for the CLI node and it worked like a charm. `ot dns config` shows it.
```
uart:~$ ot dns config
Server: [64:ff9b:0:0:0:0:808:808]:53
ResponseTimeout: 5000 ms
MaxTxAttempts: 2
RecursionDesired: yes
Done
uart:~$
```
When I resolve it works
```
uart:~$ ot dns resolve ipv6.google.com
DNS response for ipv6.google.com. - 2404:6800:4007:827:0:0:0:200e TTL:300 
Done
uart:~$ 
```
Thanks!

**Reply by jwhui** [reply]
Yes, the `srp client` and `dns config` commands are not implemented in `nRF5_SDK_for_Thread_and_Zigbee_v4.2.0_af27f76`.

To enable SRP client and DNS client in the build, include the `-DOT_SRP_CLIENT=ON` and `-DOT_DNS_CLIENT=ON` build options.

**Reply by abu-matterize** [reply]
Got it. Thank you @jwhui

**Reply by abu-matterize** [reply]
@jwhui - I've noticed one more thing. This only resolves IPv6 address and not the IPv4. Do I need to add some flag on while building to resolve IPv4 addresses?

**Reply by jwhui** [reply]
Try using `dns resolve4` to resolve IPv4 addresses.

**Reply by abu-matterize** [reply]
@jwhui - `dns resolve4` throws `Error 35: InvalidCommand`

**Reply by jwhui** [reply]
`dns resolve4` was added in #7441 here: https://github.com/openthread/openthread/blob/5daca7c684adf5124e659b0171fb1b8679e08750/src/cli/cli.cpp#L1567

It does require the following build config, but it is default enabled:
https://github.com/openthread/openthread/blob/5daca7c684adf5124e659b0171fb1b8679e08750/src/core/config/dns_client.h#L58

**Reply by abu-matterize** [reply]
Will check.

**Comment by FHFS** [comment]
I had to configure my NAT64 interface on the border router as DNS to resolve hosts.
for example with the ot cli: `ot dns resolve reddit.com fdaa:bb:1::2 53 1000 2 1`
I also needed to add a prefix and two routes, one to the well known prefix and another to the nat64 interface.

Prefix and routes on my CLI device: 
```
route add 64:ff9b:0:0:0:0::/96 s med
route add fdaa:bb:1:0:0:0::/96 s med
prefix add fd35:ad25:a99f:8c75::/64 paos med
netdata register
```
with the wpan0 interface on my border router having the ip address: `inet6 fd35:ad25:a99f:8c75:a9c8:521f:d197:eca5/64 scope global nodad
`

Hope that helps

**Reply by abu-matterize** [reply]
@FHFS - Thanks! This worked.
